### PR TITLE
Refactoring - Documentation - Fixing small bugs

### PR DESCRIPTION
**Pull request by pechouc**
# Goal of this PR

In relation to #36 and the associated [Google Doc](https://docs.google.com/document/d/17veJWmzZ9rhvFFuh_CMD9mPxrn1E8WlXoEStIADIFl8/edit?usp=sharing), this PR aims at:
* cleaning the code of the platform and standardising it as much as possible (cf. [1.1. Refactoring & Documentation](https://docs.google.com/document/d/17veJWmzZ9rhvFFuh_CMD9mPxrn1E8WlXoEStIADIFl8/edit#heading=h.g6jg4ux53ngo));
* improving the documentation of the code by reviewing docstrings and comments (cf. [1.1. Refactoring & Documentation](https://docs.google.com/document/d/17veJWmzZ9rhvFFuh_CMD9mPxrn1E8WlXoEStIADIFl8/edit#heading=h.g6jg4ux53ngo));
* fixing a few bugs (cf. [1.2. Bugs à corriger](https://docs.google.com/document/d/17veJWmzZ9rhvFFuh_CMD9mPxrn1E8WlXoEStIADIFl8/edit#heading=h.fyoboe28enx2)).

# Comment

This PR brings modifications to most Python files in the repository and I could not list all edits precisely. Most of these changes are very simple and only deal with docstrings or comments, but do not hesitate to post questions as you read through them! 

The only substantial modification concerns the callback error which appeared when switching to the `risks` map:

* The issue was that every 10 seconds, the callback updating alerts data was trying to push an alert marker to the risks map, while the latter does not have any attribute to "receive" this output. This was unlike the alerts map which had, in its `children`, an HTML div with id `live_alerts_marker` where we could inject the alert marker. 

* The proposed fix first consists in adding a similar HTML div with id `live_alerts_marker` to the risks map. This eliminates the callback error but the alert marker now appears on the risks map, which we want to avoid a priori. 

* Two additional modifications were thus made: 
  * modifying the `build_alerts_elements` in `alerts.py` so that it takes as input the style of map that the user is currently viewing and returns a void string instead of the alert marker if the user is in risks mode;
  * editing the `fetch_alert_status_metadata` callback in `main.py` so that it identifies the style of map that the user is currently viewing and passes it as argument to the `build_alerts_elements` function. 

* NB: this also allows to eliminate the `alert_button` (banner above the map) when switching to risks mode.

This renders as follows: 

(i) alert is identified and displayed in alerts mode

Capture d’écran 2020-12-26 à 15 31 15

(ii) alert attributes are hidden in risks mode, although the data is still being refreshed

Capture d’écran 2020-12-26 à 15 31 20

# Question

After a first pass on the files, I tried to automatically build an HTML documentation with [pydoc](https://docs.python.org/3/library/pydoc.html). This works quite well and renders nicely, making it (in my opinion) easier to check the documentation and understand what each function does.

Here are a few screenshots of what it looks like: 

(i) there is an index file which gives an overview of the various modules

Capture d’écran 2020-12-26 à 15 23 22

(ii) then, one can check the documentation for any Python file

Capture d’écran 2020-12-26 à 15 23 41

(iii) and even check the source code of the functions.

Capture d’écran 2020-12-26 à 15 23 48

You can make the experiment by first installing `pydoc`:

```pip install pdoc3```

And then running from the `pyro-platform/app` folder:

```pdoc --html /absolute/path/to/pyro-platform/app --output-dir /absolute/path/to/pyro-platform/app/documentation```

For the moment, I have added the `app/documentation` folder to the `.gitignore` because I feel like it is not optimal to push all the resulting HTML files in the repository. But if you find it useful, we could look for more elegant options. 

What do you think of building such a documentation? Feel free to say it's useless of course 😄

### CODE REVIEW

**Review by Akilditu — COMMENTED** [review]
Thanks a lot for this PR @pechouc !

Great refactoring job, catching up on this just now, will have a deeper look !

**Comment by Akilditu** [comment]
> i) alert attributes are hidden in risks mode, although the data is still being refreshed
> 
> Capture d’écran 2020-12-26 à 15 31 20

I'm not 100% sure we want to disable the alert banner and status when switching to risks mode. In a way it makes sense that alert informations stick to the alert mode but it could also be confusing that when switching from one map to another this 'emergency' status disapears

I guess it's a topic that we could discuss with the UX team as well.

**Comment by pechouc** [comment]
Indeed, it might make sense to indicate that there is an alert going on even in `risks` mode. 

I was thinking that since the user first loads the app in `alerts` mode, switching to `risks` meant deliberately ignoring ongoing alerts. But coming back to it, it is true that one could let the app run in `risks` mode for a long time and possibly miss alerts if they are not notified. 

However, my guess is that we cannot have the same objects indicating an alert in the two modes. Typically, the banner is associated with a user behavior: by clicking on it, the map is zoomed on the alert marker. Since this marker is not displayed on the `risks` map, leaving the banner could be misleading. 

Overall, I agree to say that this should be discussed with the UX team and I can remove this modification for now 😊

**Comment by Akilditu** [comment]
Hi @pechouc as discussed during our meetings the temporary solution would be to :

1. Have the alert CTA on both `risks` and `alerts` maps
2. Make a redirection on `alerts` once the CTA is clicked on `risks`

I guess we can close this PR once we have that done 😎

**Comment by pechouc** [comment]
After the last commit and as described by @Akilditu, the alert banner appears when the user is viewing the `risks` map (which was the case before this PR already) and a click on this banner from `risks` mode redirects towards the `alerts` map. 

In the code, this translates into three main changes:

- There are now two alert banner objects, with a separate `id` each: one for the `alerts` mode and another for the `risks` view. Each of them has its own callback in the `main.py` file, the one for `alerts` being very close to what was already written and the one for `risks` being described below. 

- As there are now two buttons that can trigger a change in the map style (the main map style button in the user selection area on the left of the page and the alert banner in `risks` mode), this requires three different callbacks, somewhat similarly to what is described [here](https://dash.plotly.com/advanced-callbacks) (section "As An Indirect Result of User Interaction"). Each button updates its own HTML "placeholder" and the change in the `children` attribute of any of these objects triggers the callback that actually operates the change in map style.

- As mentioned above, these modifications required to create two placeholders in `homepage.py` (respectively `switch_map_view_1` and -2). Another placeholder was also added to act as "source of truth" for the map style being viewed by the user, which is useful for several callbacks.

File: app/graphs.py
Comment (chloeskt) [inline app/graphs.py:1]: Hey @pechouc, thank you for this PR ! I agree with everthing. Once we merged this PR to master, we can change the calls to geojson departements and the historic fires, so as discussed in #35 

I just have a small question: for the docstring, should we use """ or ''' ? I always use """ on my end as it's described [here](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings). But we can use ''' if it's what have been chosen. However I think that we need to stay consistent throughout the repo and here I can see that we are using """ in some docstrings (for instance in `build_info_object`) and ''' in others. What do you think ?

Comment (pechouc) [reply app/graphs.py:1]: To be honest, the `'''` option has not been explicitly chosen and I did not know that there was a proper way to write these docstrings. I agree that we should stay consistent throughout the repository and for the sake of rigour, I will make the switch to `""""` in a few hours. Thanks for the heads-up!

Comment (chloeskt) [reply app/graphs.py]: Good for me 🤗

**Review by chloeskt — DISMISSED** [review]

File: app/graphs.py
Comment (pechouc) [reply app/graphs.py]: It seems that this solved the Codacy issue, nice catch!

**Review by Akilditu — APPROVED** [review]
Thanks a lot for this PR ! 

Nice improvement overall, documentation + refactoring + new alert button mechanism !

File: app/main.py
Comment (pechouc) [reply app/main.py:38]: Yes, it's a great tool! I discovered it when looking for callback techniques allowing a given event to be triggered by two different inputs, it would probably be worth using it regularly to avoid weird behaviours with the map

File: app/homepage.py
Comment (pechouc) [reply app/homepage.py]: Good idea! I have changed this in the new commit.

File: app/main.py
Comment (pechouc) [reply app/main.py:415]: Yes, exactly! I would add that it allows to figure out which input has triggered the callback _for the last time_ because I have the impression that the object stores all triggering events (at least for a given session I guess). But your understanding is accurate 😄

File: app/homepage.py
Comment (pechouc) [reply app/homepage.py]: It requires a new approval if you have a minute 😇

**Review by Akilditu — APPROVED** [review]
All good, thanks !

**Pull request merged**
